src/main/scala/is/hail/annotations/StagedRegionValueBuilder.scala

```diff
@@ -0,0 +1,174 @@
+package is.hail.annotations
+
+import is.hail.asm4s.{AsmFunction2, Code, FunctionBuilder}
+import is.hail.asm4s.Code._
+import is.hail.asm4s._
+import is.hail.expr._
+import is.hail.utils._
+import org.objectweb.asm.tree.{AbstractInsnNode, IincInsnNode}
+
+import scala.collection.generic.Growable
+import scala.reflect.ClassTag
+
+class StagedRegionValueBuilder[T] private (val fb: FunctionBuilder[AsmFunction2[T, MemoryBuffer, Long]], val rowType: Type, var region: Code[MemoryBuffer], val pOffset: Code[Long])(implicit tti: TypeInfo[T]) {
+
+  private def this(fb: FunctionBuilder[AsmFunction2[T, MemoryBuffer, Long]], rowType: Type, parent: StagedRegionValueBuilder[T])(implicit tti: TypeInfo[T]) = {
+    this(fb, rowType, parent.region, parent.currentOffset)
+  }
+  def this(fb: FunctionBuilder[AsmFunction2[T, MemoryBuffer, Long]], rowType: Type)(implicit tti: TypeInfo[T]) = {
+    this(fb, rowType, fb.getArg[MemoryBuffer](2), null)
+  }
+
+  val input: LocalRef[T] = fb.getArg[T](1)
+  var staticIdx: Int = 0
+  var idx: LocalRef[Int] = _
+  var elementsOffset: LocalRef[Long] = _
+
+  var elementsOffsetArray: LocalRef[Array[Long]] = _
+
+  rowType match {
+    case t: TStruct => elementsOffset = fb.newLocal[Long]
+    case t: TArray => {
+      elementsOffset = fb.newLocal[Long]
+      elementsOffsetArray = fb.newLocal[Array[Long]]
+      idx = fb.newLocal[Int]
+      var c = elementsOffsetArray := Code.newArray[Long](10)
+      var i = 0
+      while (i < 10) {
+        c = Code(c, elementsOffsetArray.update(i, t._elementsOffset(i)))
+        i += 1
+      }
+      fb.emit(c)
+    }
+    case _ =>
+  }
+
+  var transform: () => AsmFunction2[T, MemoryBuffer, Long] = _
+
+  val startOffset: LocalRef[Long] = fb.newLocal[Long]
+  def endOffset: Code[Long] = region.size
+
+  def currentOffset: Code[Long] = {
+    rowType match {
+      case _: TStruct => elementsOffset
+      case t: TArray => elementsOffset// + (idx.toL * t.elementByteSize)
+      case _ => startOffset
+    }
+  }
+
+  def start(): Code[Unit] = {
+    assert(!rowType.isInstanceOf[TArray])
+    rowType.fundamentalType match {
+      case _: TStruct => start(true)
+      case TBinary =>
+        assert (pOffset == null)
```

Comment:
Our code style prefers no spaces between function names and the left parentheses.